libunwind/src/UnwindCursor.hpp

```diff
@@ -3032,6 +2983,162 @@ int UnwindCursor<A, R>::stepThroughSigReturn(Registers_s390x &) {
 #endif // defined(_LIBUNWIND_CHECK_LINUX_SIGRETURN) &&
        // defined(_LIBUNWIND_TARGET_S390X)
 
+#if defined(_LIBUNWIND_CHECK_HAIKU_SIGRETURN)
+
+#if defined(B_HAIKU_32_BIT)
+typedef Elf32_Sym elf_sym;
+#define ELF_ST_TYPE ELF32_ST_TYPE
+#elif defined(B_HAIKU_64_BIT)
+typedef Elf64_Sym elf_sym;
+#define ELF_ST_TYPE ELF64_ST_TYPE
+#endif
+
+// Private syscall declared as a weak symbol to prevent ABI breaks.
+extern "C" status_t __attribute__((weak))
+_kern_read_kernel_image_symbols(image_id id, elf_sym *symbolTable,
+                                int32 *_symbolCount, char *stringTable,
+                                size_t *_stringTableSize, addr_t *_imageDelta);
+
+static addr_t signalHandlerBegin = 0;
+static addr_t signalHandlerEnd = 0;
+
+template <typename A, typename R>
+bool UnwindCursor<A, R>::setInfoForSigReturn() {
+  if (signalHandlerBegin == 0) {
+    // If we do not have the addresses yet, find them now.
+
+    // Determine if the private function is there and usable.
+    if (_kern_read_kernel_image_symbols == nullptr) {
+      signalHandlerBegin = (addr_t)-1;
+      return false;
+    }
+
+    // Get the system commpage and enumerate its symbols.
+    image_id commpageImage = -1;
+    image_info info;
+    int32 cookie = 0;
+    while (get_next_image_info(B_SYSTEM_TEAM, &cookie, &info) == B_OK) {
```

Comment:
Enumerating kernel images is risky for forward compatibility because it can be restricted to root user in future.

Reply:
Same goes to including private headers because any of them can be changed anytime. And I'd say the probability of Haiku devs moving headers around is way higher than them implementing multiuser concerns.

Reply:
Some APIs were already restricted to root user recently by @waddlesplash. I am not saying that it can't be accepted, it is fine if there currently no better way to detect signal handler function address and size.

Reply:
Probably it should just search the current team for the commpage, not the kernel team, and the "read kernel image symbols" syscall should accept image_ids of process commpages and figure out the kernel symbols from that.

Reply:
That'd require some kernel side changes, wouldn't that?

Reply:
Perhaps a `get_commpage_symbols`, or even better, `is_signal_handler_frame` and `get_signal_handler_frame_context` under `<debugger.h>` since debuggers and tools like this usually want them.

Reply:
Those sound like good ideas. What APIs do other OSes provide here, is there something we can copy directly?

Reply:
In other OSes signal handler is usually implemented in libc/ntdll.

Reply:
From what I understand for Linux, they don't provide any APIs, but instead a stable _ABI_ where magic numbers can last long.

Reply:
An example, right here in `libunwind`:

https://github.com/llvm/llvm-project/blob/1c5961c0481b9c7421d38e3141d3c5a1e6084234/libunwind/src/UnwindCursor.hpp#L2850-L2862

Reply:
Linux is a bit special here. What do the BSDs or Darwin do?

Reply:
I don't think `libunwind` supports this feature for other OSes.

Reply:
It appears GNU libunwind does. FreeBSD at least has a similar setup to Linux with a different magic number.

If what's going on here is just that the signal handler trampoline and its size need to be found, perhaps we can make it properly enumerable by `dl_iterate_phdr`? That may be more work but will probably be more portable too?

Reply:
`dl_iterate_phdr` iterates through _program headers_, which the `commpage` does not seem to have.

See also:
https://github.com/haiku/haiku/blob/a1e94e45fc773750e0d106532299bf6da8d0e513/src/libs/bsd/dl_iterate_phdr.c#L43-L47

Reply:
Yes, and my suggestion here is that we find a way to change this, and give the commpage an ELF header.

Reply:
The responsibility of kernel image enumeration has been moved to Haiku's own `libroot.so`.